src/zkevm_specs/evm/execution/sar.py

```diff
@@ -0,0 +1,172 @@
+from ...util import FQ, N_BYTES_U64, RLC
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..typing import Sequence
+
+MAX_U64 = 0xFFFFFFFFFFFFFFFF
+
+
+def sar(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    a = instruction.stack_pop()
+    shift = instruction.stack_pop()
+    b = instruction.stack_push()
+
+    (
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+        is_neg,
+    ) = __gen_witness(instruction, opcode, a, shift)
+    __check_witness(
+        instruction,
+        a,
+        shift,
+        b,
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+        is_neg,
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(2),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def __check_witness(
```

Comment:
`is_neg` should be constrained to be bool. Which is not the case here.
Also, we should check that indeed a is negative with constraints. Otherwise we could try to trample the constraints.
I'm not sure that just witnessing the `is_false` property is enough.

Reply:
I added [`instruction.constrain_bool(is_neg)`](https://github.com/scroll-tech/zkevm-specs/blob/feat/opcode-sar/src/zkevm_specs/evm/execution/sar.py#L66) here, not sure if this will work.

Reply:
Updated the code as:
```
# `is_neg` constraints
instruction.constrain_bool(is_neg)
instruction.sign_byte_lookup(
    instruction.bytes_to_fq(a.le_bytes[31:]),
    instruction.select(is_neg, FQ(255), FQ(0)),
)
```
And add `sign_byte_lookup` function to `Instruction`:
```
# `sign_byte == 0xff` if (value as i8 < 0) otherwise `sign_byte == 0`.
def sign_byte_lookup(self, value: Expression, sign_byte: Expression):
    self.fixed_lookup(FixedTableTag.SignByte, value, sign_byte, FQ(0))
```

Reply:
`is_neg` is a `LtGadget` and also updated in `check_witness` as:
```
is_neg, _ = instruction.compare(FQ(127), FQ(a.le_bytes[31]), 1)
```